src/mongo_client.ts

```diff
@@ -642,16 +642,14 @@ export class MongoClient extends TypedEventEmitter<MongoClientEvents> {
     }
 
     const session = this.startSession(options);
-    const PromiseConstructor = PromiseProvider.get() ?? Promise;
-
-    return PromiseConstructor.resolve()
-      .then(() => withSessionCallback(session))
-      .then(() => {
-        // Do not return the result of callback
-      })
-      .finally(() => {
+
+    return maybeCallback(async () => {
+      try {
+        await withSessionCallback(session);
+      } finally {
         session.endSession().catch(() => null);
-      });
+      }
```

Comment:
optional - we can remove the `.catch(() => null` in favor of try with an empty catch, like we discussed yesterday at team office

Reply:
Done!